allennlp/data/dataset.py

```diff
@@ -5,71 +5,42 @@
 
 import logging
 from collections import defaultdict
-from typing import Dict, List, Union, Iterable, Iterator, Callable
+from typing import Dict, List, Union, Iterator, Iterable
 
 import torch
-import tqdm
-from overrides import overrides
 
+from allennlp.common.checks import ConfigurationError
+from allennlp.common.util import ensure_list
 from allennlp.data.instance import Instance
 from allennlp.data.vocabulary import Vocabulary
-from allennlp.common.checks import ConfigurationError
 
 logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
 
-
-class InstanceCollection(Iterable):
+class Batch(Iterable):
     """
-    This is the abstract base class for :class:`~allennlp.data.instance.Dataset`
-    and :class:`~allennlp.data.instance.LazyDataset` objects. As its name indicates,
-    it's a collection of :class:`~allennlp.data.instance.Instance` objects
-    that can be iterated over. Depending on the subclass, the instances may be stored
-    in memory, or they may be generated (e.g. from disk) and discarded each iteration.
-    The ``Instances`` have ``Fields``, and the fields could be in an indexed or unindexed state -
-    the ``index_instances`` method indexes the data.
+    A batch of Instances. In addition to containing the instances themselves,
+    it contains helper functions for converting the data into tensors.
     """
-    def __iter__(self) -> Iterator[Instance]:
-        """
-        Returns an iterator that ranges over every Instance in the Dataset.
-        """
-        raise NotImplementedError
-
-    def index_instances(self, vocab: Vocabulary) -> None:
+    def __init__(self, instances: Iterable[Instance]) -> None:
```

Comment:
Can't you change this type annotation to `List[Instance]` given that it must always be that? Then you could remove the `ensure_list` check, particularly because you also do this check in all the `DataIterators` before creating a `Batch` anyway.